src/staking/block_validator.cpp

```diff
@@ -248,6 +313,15 @@ class BlockValidatorImpl : public AbstractBlockValidator {
     if (validation_info.GetHeight() != static_cast<blockchain::Height>(prev_block.nHeight) + 1) {
       return BlockValidationResult(Error::MISMATCHING_HEIGHT);
     }
+    std::int64_t lock_time_cutoff = prev_block.GetMedianTimePast();
+    for (const auto &tx : block.vtx) {
+      if (!IsFinalTx(*tx, validation_info.GetHeight(), lock_time_cutoff)) {
+        return BlockValidationResult(Error::NON_FINAL_TRANSACTION);
+      }
+    }
+    if (m_blockchain_behavior->GetBlockWeight(block) > m_blockchain_behavior->GetParameters().maximum_block_weight) {
```

Comment:
Why this check has to be done in `ContextualCheckBlockInternal`? It seems to not depend on context.

Reply:
I agree. I am nevertheless making sure in this pull request that our new validation produces the exact same results as the old one.

In a follow-up we can refactor and reorder. Also with respect to the document that @AM5800 and me sent around.